### PR TITLE
Add data grid view of metrics

### LINKED ISSUE #1123: Accessibility and screen reader support for metrics
We need to determine what is required (and what beyond that we want to do) for both accessibility generally and screen reader support specifically in the metrics pages. For instance, announcing new data as it appears on the screen. This was one specific area brought up by the UX board today.

### PR DESCRIPTION
**Pull request by adamint**
Adds an accessible alternative to metric graphs

As mentioned in #1123, the existing plots are inaccessible and cannot easily be made accessible. The approach taken in this pull request is:
- Show two tabs, `Graph` (this is shown by default), and `Table`
- In the table view, display a fluent data grid with two boolean settings, and varying columns based on whether we are to show a histogram or not.
- Append the content of new rows to an offscreen element that has `role=log`

There are two settings. The first toggle is whether to show the last 10 values or all values (especially in histogram views when data can update every second, the number of rows can get quite large), and by default the last 10 is selected. The second toggle is whether to only show value _changes_ (I'm looking for a better name for this setting) - i.e. we hide a metric value if it isn't different than the previous one. A good example is shown below:
image
You can see, based on a down arrow appearing in all rows, that there was a change in each visible row.

**Histograms**
image
image

Both tables have a start and end column, but histograms show P50/P90/P99 in separate columns. Values in each cell are highlighted according to the color representing that percentile in the plotly graph, and are meant to aide in column differentiation. For any given data, if there was a difference with the previous value for that percentile, an arrow icon is shown (info up if the value increased, warning down if the value decreased), and a gray right arrow if there was no change.

**Simple metrics**
The same value change indicator is shown in the value (if value is comparable) and count columns
image

**Screen reader support**
Time-based metrics are a good fit for the `log` role; unfortunately, fluent data grid does not support any live regions or roles. To work around this (& could be applied to other data grids in the future), we read out the row textual content when new rows are added. This works by:
1. Computing which indices were added (this could be extended to changed values as well)
2. Invoking a javascript function which parses the generated content for rows with those indices, and creates/adds that content to an offscreen element with `role=log`.

image

For example, as those 5 new metric values are added to visible content, we read this to screen readers:

image

Note that 1+2 and 4+5 appear on the same line because they were added nearly instantaneously. We can likely expect that, similar to navigating a data grid with a keyboard, users will navigate through the data grid header columns either initially or after logs begin being read out.

fixes #1123, future improvements could consist of allowing users to control the data refresh spacing
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1755)

### CODE REVIEW

**Comment by JamesNK** [comment]
I think you should remove column resizing and sorting in the table. Sorting doesn't make sense when only showing the 10 most recent values. And we should wait for a reason to allow resizing rather than enabling it by default.

In the tab control, How come the graph option has an icon and table doesn't?

The filters used to appear to the right of the graph on a larger display. It's stopped doing that.

Changing the text next to the switches when the value changes is unusual.

**Comment by davidfowl** [comment]
I think a more useful grid view is the latest value for each counter (rather than having to look at a grid per counter).

**Comment by adamint** [comment]
@davidfowl I don't understand what you mean - are you still talking about the view of a specific metric

**Comment by adamint** [comment]
> I think you should remove column resizing and sorting in the table. Sorting doesn't make sense when only showing the 10 most recent values. And we should wait for a reason to allow resizing rather than enabling it by default.

Good idea, I've made this change. 

> In the tab control, How come the graph option has an icon and table doesn't?

That was an intentional decision to draw attention to the graph, and not the table, but I can add an icon to the table too!

> The filters used to appear to the right of the graph on a larger display. It's stopped doing that.

> Changing the text next to the switches when the value changes is unusual. The switches are to the left of text. Switches normally appear to the right (at least that it what I saw when looking through Windows UI for switches). It's inconsistent with the "Show count" switch on histogram graphs.

Good call, changed. 
image

> Histogram + show count + table doesn't display a good value.

That's due to Histogram's `ToString`, which is currently being displayed; I've removed the value column in the case where show counts is enabled for a histogram

**Comment by tlmii** [comment]
1. Playing with it for a bit with different metrics, I don't particularly like having the "Show the latest 10 values" disabled when there aren't yet 10 items. I understand the initial inclination to do so, but I think it just adds friction to the user if they know there will be more than 10 and they have to wait for all 10 to appear and _then_ toggle it. It also gets disabled again as data disappears.
2. I think we should attempt to layout the page such that the grid doesn't grow as data gets added. It looks like right now its got a max-height of `60vh`. But for the first while, as the data gets added, the toggles and the filters all get pushed down the page.
3. The data disappearing from the grid (as it scrolls off the left side of the chart) is a bit awkward and I worry a bit about how well that will work from an accessibility/screen reader perspective. Don't have any real good suggestions at the moment, just wanted to call it out to see how others felt.
4. The gray right arrow (I believe indicating the value didn't change) and the little chart icon in the start column don't meet contrast in the dark background (neither does the disabled text of the switches, but disabled controls don't need to meet contrast)

**Comment by JamesNK** [comment]
The order of data is oldest at the top, newest at the bottom. I know that is the order of structured logs and traces (chosen to make those pages feel like a terminal) but it doesn't feel right for this metrics table. I don't know how to articulate why.

**Comment by JamesNK** [comment]
Need margin on the bottom of the table to create space between Filters header:

image

File: src/Aspire.Dashboard/Components/Controls/MetricTable.razor
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/MetricTable.razor:9]: Why is virtualize false? There is a noticeable delay when switching between 10 latest rows and all rows. The delay is from adding all the new rows to the table. That's not a problem with virtualization.

Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/MetricTable.razor:9]: Blazor errors if the duration is set to 30 minutes or greater, apps have been running that long, and you switch to view all data.

Virtualization is definitely required.

**Comment by JamesNK** [comment]
Add 5 pixels between icon and text in tab:

image

And add 5 pixels between tabs and content below.

**Comment by JamesNK** [comment]
Graph has the wrong size when switching between metrics as table and then switching back to the graph:

metrics-table-switch

File: src/Aspire.Dashboard/Components/Controls/MetricTable.razor
Comment (adamint) [reply src/Aspire.Dashboard/Components/Controls/MetricTable.razor:9]: If we flip the order so that new data points are added to the top of the table instead of the bottom, re-adding virtualization is easy. The current approach for reading to screen readers is to obtain the text content of each cell in the latest row, which is not possible if the row is virtualized away. 

I'll change the order, which will take care of this & your previous comment.

**Comment by adamint** [comment]
> Playing with it for a bit with different metrics, I don't particularly like having the "Show the latest 10 values" disabled when there aren't yet 10 items. I understand the initial inclination to do so, but I think it just adds friction to the user if they know there will be more than 10 and they have to wait for all 10 to appear and then toggle it. It also gets disabled again as data disappears.

Changed to always being enabled.

> I think we should attempt to layout the page such that the grid doesn't grow as data gets added. It looks like right now its got a max-height of 60vh. But for the first while, as the data gets added, the toggles and the filters all get pushed down the page.

I set the height to 40vh now!

> The data disappearing from the grid (as it scrolls off the left side of the chart) is a bit awkward and I worry a bit about how well that will work from an accessibility/screen reader perspective. Don't have any real good suggestions at the moment, just wanted to call it out to see how others felt.

Sorry, what do you mean by this?

> The gray right arrow (I believe indicating the value didn't change) and the little chart icon in the start column don't meet contrast in the dark background (neither does the disabled text of the switches, but disabled controls don't need to meet contrast)

This is a problem more generally with the --info, --warning, and --success colors. I brightened all of them

> The order of data is oldest at the top, newest at the bottom. I know that is the order of structured logs and traces (chosen to make those pages feel like a terminal) but it doesn't feel right for this metrics table. I don't know how to articulate why.

Changed.

> Need margin on the bottom of the table to create space between Filters header:

Added margin

> There should be a consistent space between switch text and switch

Done

> Add 5 pixels between icon and text in tab

5px appears too spaced out, but I added 2px.

> And add 5 pixels between tabs and content below.

Done

> Graph has the wrong size when switching between metrics as table and then switching back to the graph:

There seems to be an issue with plotly here, as the svg width is being set to 700 instead of 650 somehow? Regardless, an easy workaround is to hide the plot overflow.

**Comment by adamint** [comment]
image
image

**Comment by adamint** [comment]
Colors were changed just enough to meet the 3:1 required contrast

**Comment by tlmii** [comment]
>> The data disappearing from the grid (as it scrolls off the left side of the chart) is a bit awkward and I worry a bit about how well that will work from an accessibility/screen reader perspective. Don't have any real good suggestions at the moment, just wanted to call it out to see how others felt.

> Sorry, what do you mean by this?

Since the table only shows the data currently displayed in the chart, it doesn't just grow in length, it can also shrink. For instance if you look at the requests metric, hit the page a few times, there will be a bunch of data in the table. Now wait 5 min (or whatever the selected time range is) and most of the data will disappear from the table (at the same time that it scrolls off the left side of the chart).

It's an odd user experience to have rows disappear from a table of data. Beyond that, I just wonder about accessibility (for instance, if a screen reader is trying to read the data (which takes much longer than visually seeing it) and then it  scrolls off. I guess the answer there could be to use a longer time frame. I just want to make sure we're aware of the implications.

**Comment by JamesNK** [comment]
From call:

image

**Comment by JamesNK** [comment]
I fixed plotly chart. The problem is the chart is initialized when switching metric, but it's not rendered if the table is selected, and so can't figure out what size it should be. Fix is to be explicit.

**Comment by adamint** [comment]
image

- Made arrows filled, added tooltip
- Removed latest values switch
- Made sure start/ends add up when Show only value changes is selected
- Fixed filter popups not showing up (TIL that even though there could only be one instance of it, we could still run into id issues)
- Made scroll bar auto
- Added 1px to histogram value underlines

I might be misunderstanding, but after putting some breakpoints, I think the small window where the old data appears with the updated data grid might be a fluentui-blazor issue? At the point where that is occurring, the bound data is empty.


After taking a look, I think it might be a good idea to do pause & data window selection together, later, as it will also affect the chart 😄

**Review by tlmii — APPROVED** [review]

**Comment by JamesNK** [comment]
Latest source code errors for me:

```

2024-01-26T07:12:36.9023688 warn: Microsoft.AspNetCore.Components.Server.Circuits.RemoteRenderer[100]
      Unhandled exception rendering component: Failed to fetch dynamically imported module: http://localhost:15888/_content/Aspire.Dashboard/Components/Controls/MetricTable.razor.js
TypeError: Failed to fetch dynamically imported module: http://localhost:15888/_content/Aspire.Dashboard/Components/Controls/MetricTable.razor.js
      Microsoft.JSInterop.JSException: Failed to fetch dynamically imported module: http://localhost:15888/_content/Aspire.Dashboard/Components/Controls/MetricTable.razor.js
TypeError: Failed to fetch dynamically imported module: http://localhost:15888/_content/Aspire.Dashboard/Components/Controls/MetricTable.razor.js
         at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
         at Aspire.Dashboard.Components.Controls.MetricTable.OnAfterRenderAsync(Boolean firstRender) in C:\Development\Source\aspire\src\Aspire.Dashboard\Components\Controls\MetricTable.razor.cs:line 38
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)
2024-01-26T07:12:36.9214161 fail: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[111]
      Unhandled exception in circuit 'hJrn4jK0KXZZLoY1eO8to3-G6pbi7-WauTnDtOOQ9JE'.
      Microsoft.JSInterop.JSException: Failed to fetch dynamically imported module: http://localhost:15888/_content/Aspire.Dashboard/Components/Controls/MetricTable.razor.js
TypeError: Failed to fetch dynamically imported module: http://localhost:15888/_content/Aspire.Dashboard/Components/Controls/MetricTable.razor.js
         at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
         at Aspire.Dashboard.Components.Controls.MetricTable.OnAfterRenderAsync(Boolean firstRender) in C:\Development\Source\aspire\src\Aspire.Dashboard\Components\Controls\MetricTable.razor.cs:line 38
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)
```

File: src/Aspire.Dashboard/Components/Controls/MetricTable.razor.cs
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/MetricTable.razor.cs:27]:
```suggestion
    public required IJSRuntime JS { get; init; }
```

File: src/Aspire.Dashboard/Components/Controls/MetricTable.razor.cs
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/MetricTable.razor.cs:38]: Add async disable to component and dispose module.

**Comment by JamesNK** [comment]
The table doesn't have live updates when showing only unchanged values:

metric-table-live-updates

The end date of the latest value doesn't update.

**Comment by JamesNK** [comment]
I pushed an update to fix the flash of bad content shown here: https://github.com/dotnet/aspire/pull/1755#issuecomment-1909010699

The problem appears come from the UI updating the columns of the data grid before the data grid updates its data. I tried to find a clean way to fix it but ended up just duplicating the table 3 times. That forces the data grid to be recreated with new data when switching between instruments. It would be great if you could reduce the duplication between the 3 grids.

**Comment by JamesNK** [comment]
After my change the end date with "Only show value updates" updates. Most of the time it works, but on some tables I saw two rows with the end date updated to now:

metric-table-live-updates-2

**Comment by JamesNK** [comment]
I think some unit tests are required for calculating these metrics. The code is complicated and we could start fixing bugs in one place and create new ones in another.

File: src/Aspire.Dashboard/Components/Controls/MetricTable.razor.cs
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/MetricTable.razor.cs:57]:
```suggestion
```

**Comment by JamesNK** [comment]
I still see multiple end dates update:

metrics-table-end-date

**Comment by JamesNK** [comment]
What do you think about making the selected view (chart vs table) a query string parameter? It would be persisted and restored from state.

I think it would be an improvement but I'm happy for it to be a separate issue that comes later.

**Comment by kvenkatrajan** [comment]
> making the selected view (chart vs table) a query string parameter

I have opened issue: #1989 for tracking

**Comment by adamint** [comment]
@tlmii could I get a re-review

**Comment by JamesNK** [comment]
> This is ok, we are receiving multiple metrics whose end dates are updating

I'm confused what this means. In this comment's image, isn't the chart showing a value of 1? What does the updated value for 0 mean?

The two rows that are being updated, are they different dimensions? In the chart the dimensions are merged together to show a single value. The table should do the same thing.

**Comment by adamint** [comment]
> You can remove the icon

I'd like to keep it, it is the indicator that there's a tooltip for that column.

**Comment by JamesNK** [comment]
> I'd like to keep it, it is the indicator that there's a tooltip for that column.

Ok.

> We should consider customer feedback on this before we look at aggregating the dimensions. "Aggregate by Dimension" could be an option we add based on feedback. I can see for screensupport users needing to tune into individual events and aggregated events both.

I think that's the wrong direction. We should consider customer feedback before deciding **not** to aggregate the dimensions.

The default behavior for the table should be to display the same data as the chart. The chart sums dimensions together so the table should follow that. The current difference in behavior between the two views will be confusing. For example, someone viewing the active connections chart will see there are 10 active connections and then view the table and see the top value is 1 (because other values are in other dimensions). They would expect the table to report the latest value of 10.

I'm not against adding an option to break down values by dimension, but I think it should wait for customer feedback. And enabling it should apply to both the chart and the table, i.e. setting it to true would change the chart to display a different line for each dimension. That would keep the UI behavior between the chart and table in sync.

**Comment by adamint** [comment]
We don’t need to keep parity between the table and the chart - they fundamentally serve different audiences

**Comment by JamesNK** [comment]
No one has ever said they're confused by the chart showing a sum of all dimensions. Why would it be different for the table? And knowing what the filters are is possible today. They're on the screen in the filters section.

If anything, having aggregate values seems like it would be more friendly to screen readers. The dimensions don't need to be read out for every row. They're the same. The reader just needs the date and value. It would be much faster. And I believe there would only be one row updating at a time. That would be faster still.

Here is another problem with the current UI that would be solved by aggregating them together:

* Imagine a metric has 10 dimensions
* One of those dimensions is constantly increasing and decreasing
* Because the table is ordered by start date, all the user sees in the grid is that one dimension changing value.
* The other 9 dimensions have had the same value for a long time, so the start date is far in the past so aren't visible. The user would have no idea where to find them in the table.

**Comment by kvenkatrajan** [comment]
> The default behavior for the table should be to display the same data as the chart. 
While I do like consistency in views, this table display was intended to display raw data so users could determine deltas in updated dimensions (similar to otel data being ingested in Azure data explorer : https://learn.microsoft.com/en-us/azure/data-explorer/open-telemetry-connector?tabs=command-line). 

That being said, we took a step back to evaluate what information we were attempting to display and realized we had taken the approach to display events  specific data(which is better suited for the trace approach) rather than a histogram specific data points (more suited for metrics). We agreed that this could change the measurements on %ile distribution when the dimension data is viewed individually and could cause confusion between the views, more so now, since they will be displayed side by side in the UI.

Adam is working on the changes & Ill open another issue to track creating an option to break down values by dimension - most likely post GA

Thanks everyone for weighing in to ensure we have a great user experience.

**Comment by adamint** [comment]
I see your point James, and now that I understand, I agree with you. I've changed it such that metric values with the same end date get collapsed

**Comment by JamesNK** [comment]
This tooltip doesn't make sense because the reported value includes other dimensions:

image

It has a server port of 16031 but the total includes values from other server ports.

**Comment by JamesNK** [comment]
Histogram table count doesn't show the total count. It just repeats 1. That's inconsistent with other counters and with the chart.

Compare the table with the chart which does show the actual increasing count:

show-count

When I looked at all the code in the metric table code behind, it feels like there is a lot of duplication. They should be showing the same data but in a different view. If that was happening, then this issue wouldn't exist.

Couldn't the table view reuse the same logic as the chart page? The chart page is calculating what a metric should be each second. The table could then reuse that. That would avoid bugs like this one.

Having one source of truth for a metrics value at each point in time can then be used by both views, and each view modifies it when needed (i.e. the table collapsing duplicate rows if enabled).

**Comment by JamesNK** [comment]
A small variation in when metrics are exported and received by the dashboard means they aren't all received the exact same time. See:

end-date

The metrics chart starts displaying data a second in the past to avoid this: https://github.com/dotnet/aspire/blob/55c192136fd3c547e7ad1feac60858588fc0f23a/src/Aspire.Dashboard/Components/Controls/PlotlyChart.razor.cs#L88-L91

**Comment by kvenkatrajan** [comment]
Thanks I am hoping these will be the last bit of changes before we merge. Any further defects can be tracked as new issues. To summarize the changes requested are:

- Histogram table count doesn't show the total count. It just repeats 1. 
- The "Show count" setting isn't preserved when switching back and forth between the views. -> I would file this as a separate bug 
- A small variation in when metrics are exported and received by the dashboard means they aren't all received the exact same time. 
- Having one source of truth for a metrics value at each point in time can then be used by both views -> We will create a new issue to track. This would be a larger update since the graph view regenerates data model every update, in contrast the table updates the existing model on presence of data in order to figure out what has changed for screen reader support.

**Comment by kvenkatrajan** [comment]
Created issue to track - Have one source of truth for a metrics value at each point in time can then be used by both views: https://github.com/dotnet/aspire/issues/2104

**Comment by adamint** [comment]
> Having one source of truth for a metrics value at each point in time can then be used by both views, and each view modifies it when needed (i.e. the table collapsing duplicate rows if enabled, adding a column to indicate whether the row value is increasing or decreasing, etc).

This is a good idea; I had initially taken a different approach because the table had to be aware of what points had changed, but there is a lot duplicated.

**Comment by adamint** [comment]
@JamesNK I cannot reproduce the multiple values updating, but now values within the last second are filtered out like the plot.

**Comment by JamesNK** [comment]
Show count isn't updating.

It's over 1000 in the graph but stopped counting at 329 in the table:

show-count-incorrect

**Comment by JamesNK** [comment]
The histogram chart and table are now in sync when "Show count" is changed, but the actual toggle isn't in sync.

show-count-toggle-not-updating

File: src/Aspire.Dashboard/Components/Controls/MetricTable.razor.cs
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/MetricTable.razor.cs:69]: Is this the 1 second delay? Could you add a comment explaining why to do it.

nit: DateTime supports < and > operators. Those are usually used rather than CompareTo

**Comment by kvenkatrajan** [comment]
@JamesNK  please confirm if anything else is required here. There are bugs filed for the remaining items. Since this is a larger PR would like to avoid pushing it towards the end of preview.

**Comment by timheuer** [comment]
Any chance we can provide padding on the precision measurements? the pixel person in my is hurting when I see this:
image

where because 0.05 and 0.025 drop the conssitency of the width and in a left-aligned view (especially with icons) it looks kinda wonky.  Pad all numbers to .000 precision? so this would have been 0.050 and 0.025

**Comment by adamint** [comment]
@tlmii padded histogram values to 3 decimal points
image

I refactored `PlotlyChart` and `MetricTable` to use the same data source (thanks @JamesNK), both derive from `ChartBase` and implement `OnChartUpdated` to update their respective views.

image

The table view thus now always shows the same values as the plot. 

Additionally, the view you have currently selected is saved in the query

**Comment by tlmii** [comment]
I think that was @timheuer but I agree it looks better, nice job!

**Comment by samsp-msft** [comment]
Can we change this to "HH:mm:ss" as we don't need the date (its only the last 5 minutes). I was going to suggest the milliseconds, but as the metrics get aggregated by default on a second basis, I am not sure how much use that is?

**Comment by JamesNK** [comment]
I got the latest changes in this branch and ran the app. Unfortunately all the metric data tables are empty:

image

I get this for all counters and options.

**Comment by JamesNK** [comment]
> Any chance we can provide padding on the precision measurements? the pixel person in my is hurting when I see this: image
> 
> where because 0.05 and 0.025 drop the conssitency of the width and in a left-aligned view (especially with icons) it looks kinda wonky. Pad all numbers to .000 precision? so this would have been 0.050 and 0.025

That would be an improvement in this particular example. However, the problem with giving all numbers .000 precision is not all numbers will have decimal points.

It will be noisy if the data was 1000, 1001, 1002, etc, but they were displayed in the table as 1000.000, 1001.000, 1002.000, etc, instead.

I don't think it should be done everywhere. We can do this smarter. For example, histograms have a known set of values (the buckets). We could look at the histogram bucket with the most precision and then give all values that level of precision. And for regular charts, if we have any non-whole numbers then include decimal points then.

Those are more complicated. I recommend undoing the change to add precision everywhere and creating a follow up issue to make an improvement that works everywhere.

**Comment by JamesNK** [comment]
> @tlmii padded histogram values to 3 decimal points image

Can you look into why the table columns have uneven sizes. P50, P90, P99 should be evenly sized.

**Comment by adamint** [comment]
> got the latest changes in this branch and ran the app. Unfortunately all the metric data tables are empty

I do not see this

**Comment by JamesNK** [comment]
I did a git pull on the branch, verified my local SHA was the same as the PR's server SHA, did a git clean, and rebuilt the app. I don't see any data:

metrics-no-data

**Comment by JamesNK** [comment]
@adamint I found the issue. `UpdateMetrics` is testing xValues against DateTime.UtcNow:

```
if (DateTime.UtcNow - xValue < TimeSpan.FromSeconds(2))
{
    continue;
}
```

These values are already local. All values always pass on your timezone (after UTC) and always fail in my timezone (before UTC).

**Comment by JamesNK** [comment]
@adamint It's late for you. If you'd like I can make changes based on testing rather than waiting. That would speed things up.

**Comment by adamint** [comment]
Ah, I see now… I saw UtcNow was being used elsewhere, so erroneously used it here too

> @adamint It's late for you. If you'd like I can make changes based on testing rather than waiting. That would speed things up.

Yes, thank you!

**Review by JamesNK — APPROVED** [review]

**Comment by JamesNK** [comment]
I made some changes - [`7991177` (#1755)](https://github.com/dotnet/aspire/pull/1755/commits/799117785d21b1d145b6a839348a96da4d1e1674)

The only odd issue remaining that I noticed is there is slightly weird data added when the next row is added to the chart. The previous row time changes by a second, and sometimes an extra row briefly appears.

**Comment by adamint** [comment]
Thank you again james! 😄

**Pull request merged**
